### DISCUSSION TITLE
Example of generic connection pool

### DISCUSSION THREAD

**Discussion by skoky** [diesel-rs/diesel#3935] [Q&A] [answered]
Do we have any example of a generic connection pool that accepts Postres and MySql urls? I don't want to specify the database driver in code, but rather as parameter with url like postgres://... or mysql://... and have a code that is generic and inserts data to the database. What I can't find is a database trait that I could return from function 'establish_connection_pool(url: String)' and use it for inserts and updates. Thanks

**Comment by weiznich** [comment]
It's hard to answer such super generic questions. Also it's always a good idea to [search existing questions](https://github.com/diesel-rs/diesel/discussions?discussions_q=is%3Aopen+generic+connection) before opening a new discussion thread (or at least link to existing questions from your new thread and explain why you did not choose the solution presented there.)

Given the minimal amount of details in this question I can only point you to [`#[derive(MultiConnection)]`](https://docs.diesel.rs/2.1.x/diesel/derive.MultiConnection.html) as that was already done by the existing answers.

**Reply by skoky** [reply]
ok, let me be more specific

```

pub fn establish_connection_pool(database_url: &str) -> GenericDbPool {
    let manager = ConnectionManager::<GenericDbPool>::new(database_url);
    r2d2::Pool::builder()
        .build(manager)
        .expect("Failed to create pool.")
}

fn main() {
    let mysql = establish_connection_pool("mysql://myhost");
    let postgres = establish_connection_pool("postgres://myhost");

    mysql.select("select * from mytable");
    postgres.select("select * from mytable");
}
```

Can you help me fix the code, the  `establish_connection_pool` function that created connection to both Mysql and Postgres?

Thanks

**Reply by skoky** [reply]
I am looking at the example in [#[derive(MultiConnection)]](https://docs.diesel.rs/2.1.x/diesel/derive.MultiConnection.html). It looks similar to what I need, but I need connection pool. Not just a single connection to database.

**Reply by weiznich** [reply, marked as answer]
Well, if you want to have a pool you need to create a connection pool with whatever type you used `#[derive(MultiConnection)]` onto. It's supported by the built-in r2d2 pooling support as well, so that type will just work there as any `Connection` type from diesel.
